### ISSUE TITLE
[BUG]: Cannot use column as input to mapping column

### ISSUE THREAD

**Issue by camer314** [dotnet/spark#201]
In pySpark if i want to create a new column based off a dictionary I can do the following:

```python
    lookup_table1 = { 'A': 100, 'B': 200 }
    mapping_col = create_map([lit(x) for x in chain(*lookup_table1.value.items())])
    df = df.withColumn("map_test", mapping_expr.getItem(col("col1")))  
```
In other words, I can transform an entire column using the values of an input column as keys into my mapping column.

Trying to reproduce this I have the following C#:

```C#
            lookup_table1 = new Dictionary<string,int>()
            {
                { "A", 100 },
                { "B", 200 },
            };

            var cols = new List<Column>();
            
            foreach(var kv in lookup_table1)
            {
                cols.Add(Functions.Lit(kv.Key));
                cols.Add(Functions.Lit(kv.Value));
            }

            var mapping_col = Functions.Map(cols.ToArray());
            df = df.WithColumn("map_test",mapping_col.GetItem(Functions.Col("col1")));
```

This results in the error:

[Error] [JvmBridge] java.lang.RuntimeException: **Unsupported literal type class org.apache.spark.sql.Column col1**

Should i be able to replicate this functionality and if so what am i doing wrong?

**Comment by imback82** [comment]
@suhsteve this should be also related to your work item #202.

**Comment by imback82** [comment]
@camer314 can you try
```C#
df = df.WithColumn("map_test",mapping_col.GetItem("col1")); // not Functions.Col("col1")
```

The following works for me:
```C#
var df = spark.Range(0, 5);
var lookup_table1 = new Dictionary<string, int>()
{
    { "A", 100 },
    { "B", 200 },
};

var cols = new List<Column>();

foreach (var kv in lookup_table1)
{
    cols.Add(Functions.Lit(kv.Key));
    cols.Add(Functions.Lit(kv.Value));
}

var mapping_col = Functions.Map(cols.ToArray());
df = df.WithColumn("map_test", mapping_col.GetItem("A"));
df.Show();
```

Prints:
```
+---+--------+
| id|map_test|
+---+--------+
|  0|     100|
|  1|     100|
|  2|     100|
|  3|     100|
|  4|     100|
+---+--------+
```

**Comment by camer314** [comment]
Sorry, my original example is not complete. Here is a better one:

```python
from pyspark.sql.functions import col, create_map, lit
from itertools import chain

df = spark.range(0,6)
df = df.withColumn('keys', col('id') % 3)

lookup = { 0: 100, 1: 200, 2: 300 }
mapping_col = create_map([lit(x) for x in chain(*lookup.items())])

# values column uses keys column values as key into lookup dictionary
df = df.withColumn("values", mapping_col.getItem(col('keys')))

df.show()
```

Produces the following. Note the values column. The value of 'keys' is used as a key into the dictionary...

```
+---+----+------+
| id|keys|values|
+---+----+------+
|  0|   0|   100|
|  1|   1|   200|
|  2|   2|   300|
|  3|   0|   100|
|  4|   1|   200|
|  5|   2|   300|
+---+----+------+
```
Meanwhile, the exact conversion to C# does not work as expected:

```C#
    var df = _spark.Range(5);
    df = df.WithColumn("keys", Col("id") % 3);

    var lookup = new Dictionary<int, int>()
    {
        { 0, 100 },
        { 1, 200 },
        { 2, 300 },
    };

    var cols = new List<Column>();

    foreach (var kv in lookup)
    {
        cols.Add(Functions.Lit(kv.Key));
        cols.Add(Functions.Lit(kv.Value));
    }

    var mapping_col = Functions.Map(cols.ToArray());
    df = df.WithColumn("values", mapping_col.GetItem(Col("keys")));
    df.Show();
```

Produces:

```
[JvmBridge] java.lang.RuntimeException: Unsupported literal type class org.apache.spark.sql.Column keys
```

While this change produces NULLs:

```C#
df = df.WithColumn("values", mapping_col.GetItem("keys"));
```
```
+---+----+------+
| id|keys|values|
+---+----+------+
|  0|   0|  null|
|  1|   1|  null|
|  2|   2|  null|
|  3|   0|  null|
|  4|   1|  null|
+---+----+------+
```

**Comment by imback82** [comment]
Scala behaves same as C#:
```Scala
scala> val df = spark.range(5)
df: org.apache.spark.sql.Dataset[Long] = [id: bigint]

scala> val df2 = df.withColumn("keys", col("id") % 3)
df2: org.apache.spark.sql.DataFrame = [id: bigint, keys: bigint]

scala> val mapping_coll = map(lit(0), lit(100), lit(1), lit(200), lit(2), lit(300))
mapping_coll: org.apache.spark.sql.Column = map(0, 100, 1, 200, 2, 300)

scala> df2.withColumn("values", mapping_coll.getItem(col("keys"))).show
java.lang.RuntimeException: Unsupported literal type class org.apache.spark.sql.Column keys
  at org.apache.spark.sql.catalyst.expressions.Literal$.apply(literals.scala:78)
  at org.apache.spark.sql.Column.getItem(Column.scala:856)
  ... 49 elided

scala> df2.withColumn("values", mapping_coll.getItem("keys")).show
+---+----+------+
| id|keys|values|
+---+----+------+
|  0|   0|  null|
|  1|   1|  null|
|  2|   2|  null|
|  3|   0|  null|
|  4|   1|  null|
+---+----+------+

```

Let me check what's different about Python.

**Comment by imback82** [comment]
OK. So python behavior is different. In Scala, you have to do the following:
```Scala
// You need to call apply() instead of getItem() on MapType Column.
scala> df2.withColumn("values", mapping_col(col("keys"))).show 
+---+----+------+
| id|keys|values|
+---+----+------+
|  0|   0|   100|
|  1|   1|   200|
|  2|   2|   300|
|  3|   0|   100|
|  4|   1|   200|
+---+----+------+

```
Unfortunately, C# is missing `Apply()`. I will create a PR for this. I quickly tested this and works as expected:
```C#
var mapping_col = Map(cols.ToArray());
df.WithColumn("values", mapping_col.Apply(Col("keys"))).Show();
+---+----+------+
| id|keys|values|
+---+----+------+
|  0|   0|   100|
|  1|   1|   200|
|  2|   2|   300|
|  3|   0|   100|
|  4|   1|   200|
+---+----+------+
```

**Comment by camer314** [comment]
Cool thanks

**Comment by imback82** [comment]
Closed by #323

**Comment by imback82** [comment]
@camer314 I also fixe the behavior of Python to be consistent with Scala in Spark 3.0: https://github.com/apache/spark/pull/26351.
